schema.py

```diff
@@ -266,22 +267,34 @@ def is_valid(self, data):
         else:
             return True
 
+    def _raise_error(self, exception, message, *error_list):
+        """Raises exception with custom schema name prepended to exception
+        message and feeds additional arguments into schema error class.
+        """
+        if not self._name and not error_list:
+            raise exception("%s" % (message,))
+        elif not self._name and error_list:
+            raise exception("%s" % (message,), error_list[0])
+        elif self._name and error_list:
+            raise exception("%r %s" % (self._name, message,),
+                            error_list[0])
+
     def validate(self, data):
         Schema = self.__class__
-        s = self._schema
-        e = self._error
         i = self._ignore_extra_keys
-        flavor = _priority(s)
+        def format_error(e): return e.format(data) if e else None
```

Comment:
I think this logic can be a part of `_raise_error`.

Reply:
Taken from my above answer:

I looked into including the `format_error()` function into `_raise_error()`. Unfortunately there are some instances where `self._error` is accessed directly without being formatted (via `self._error.format(data)`) or set to `None` and other instances where the `*error_list` argument is given a value where `self._error` is neither directly accessed nor formatted.

One line 380 to 381:
`self._raise_error(SchemaError, message,
                                  format_error(self._error))`
A `SchemaError` is raised that formats or sets `self._error` to `None`

On line 395:
`self._raise_error(SchemaError, message, self._error)`
Another `SchemaError` is raised, except `self._error` is accessed directly.

There are also other instances where various schema errors access `self._error` both directly and after formatting it as well.

On line 376-377:
`self._raise_error(SchemaError, [None] + x.autos,
                                  [self._error] + x.errors)`
`self._error` is neither formatted, set to none, nor just simply accessed directly.

So for the above reasons I didn't want `_raise_error()` to perform `format_error()` on the argument given for `*error_list` every time. And since `format_error()` is both applied and not applied for the same type of schema errors, I can't easily set `format_error()` to only be executed based on the exception that's raised.

The only solution I can think of at the moment is to have `self._raise_error()` take in a fourth argument that "flags" whether `format_error()` should be executed on the `*error_list` argument.

It's at this point that I figured, for now at least, it's simpler to leave `format_error()` external to `self._reaise_error()`